### PR TITLE
Identity Metadata Repository setup

### PR DESCRIPTION
**Pull request by lmuntaner**
Main motivation is to set up the module to manage the anchor metadata.

<hr/><details><summary>🟡 Some screens were changed</summary></details>

### CODE REVIEW

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
+import { MetadataMapV2 } from "$generated/internet_identity_types";
+import type { AuthenticatedConnection } from "$src/utils/iiConnection";
+
+export type AnchorMetadata = {
+  recoveryPageShownTimestampMillis?: number;
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:5]: Doesn't this need to be `bigint`?

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+};
+
+const RECOVERY_PAGE_SHOW_TIMESTAMP_MILLIS = "recoveryPageShownTimestampMillis";
+const METADATA_FETCH_TIMEOUT = 1_000;
+const METADATA_FETCH_RETRIES = 10;
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:10]: 10 seems like a lot, no?

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        this.rawMetadata === "error" ||
+        this.rawMetadata === "not-loaded"
+      ) {
+        // Retry in case of error or not loaded.
+        void this.loadMetadata();
+      } else {
+        break;
+      }
+    }
+    if (
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:89]: I'd suggest to extract a helper for this condition.

File: src/frontend/src/utils/iiConnection.ts [hunk 4]
```diff
@@ -511,6 +521,34 @@ export class AuthenticatedConnection extends Connection {
     await actor.remove(this.userNumber, publicKey);
   };
 
+  getIdentityInfo = async (): Promise<
```
Comment (frederikrothenberger) [inline src/frontend/src/utils/iiConnection.ts:524]: A client should only use the `getAnchorMetadata`, `setPartialMetadata` and `commitMetadata` functions. `getIdentityInfo` and `setIdentityMetadata` should be `private` then, no?

**Review by frederikrothenberger — COMMENTED** [review]
Yes, this is the right direction. Not sure about the complexity of retries though...
It makes the front-end more robust, but the PR #2517 didn't have it either?

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 1 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts:5]: I will get it from the browser with `Date.now()` which returns a `number`.

From chatGPT:

> Date.now() will return Number.MAX_SAFE_INTEGER sometime around the year 287,586

If we were using nanoseconds, then yes. But with milliseconds it should be enough.

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 3 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts:89]: makes sense

File: src/frontend/src/utils/iiConnection.ts (hunk 4 above)
Comment (lmuntaner) [reply src/frontend/src/utils/iiConnection.ts:524]: Then the repository won't be able to use it...

I could pass the getter and setter as functions only.

**Comment by lmuntaner** [comment]
> Yes, this is the right direction. Not sure about the complexity of retries though... It makes the front-end more robust, but the PR #2517 didn't have it either?

I was thinking and yes, I think we don't need retries. Even more, I think I want to swallow the errors and log a warning only.

The metadata is not crucial, we can find sensible defaults in case something isn't working as expected. Moreover, with the imperative code, I don't want an error to get metadata to disrupt a user flow.

@frederikrothenberger What do you think if I remove all the throws and instead swallow or return `undefined` or `false` accordingly?

**Comment by frederikrothenberger** [comment]
> > Yes, this is the right direction. Not sure about the complexity of retries though... It makes the front-end more robust, but the PR #2517 didn't have it either?
> 
> I was thinking and yes, I think we don't need retries. Even more, I think I want to swallow the errors and log a warning only.
> 
> The metadata is not crucial, we can find sensible defaults in case something isn't working as expected. Moreover, with the imperative code, I don't want an error to get metadata to disrupt a user flow.
> 
> @frederikrothenberger What do you think if I remove all the throws and instead swallow or return `undefined` or `false` accordingly?

Yes, a warning should be sufficient. 👍

File: src/frontend/src/flows/authorize/index.ts [hunk 5]
```diff
@@ -212,6 +212,7 @@ const authenticate = async (
   const derivationOrigin =
     authContext.authRequest.derivationOrigin ?? authContext.requestOrigin;
 
+  // TODO: Commit state
```
Comment (lmuntaner) [inline src/frontend/src/flows/authorize/index.ts:215]: To be done in another PR, but I want to remember exactly where to put it.

File: src/frontend/src/flows/manage/index.ts [hunk 6]
```diff
@@ -250,6 +250,7 @@ export const renderManage = async ({
   for (;;) {
     let anchorInfo: IdentityAnchorInfo;
     try {
+      // TODO: commit state
```
Comment (lmuntaner) [inline src/frontend/src/flows/manage/index.ts:253]: To be done in another PR, but I want to remember exactly where to put it.

**Comment by lmuntaner** [comment]
@frederikrothenberger ready for review

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 7]
```diff
@@ -53,114 +86,107 @@ export class AnchorMetadataRepository {
...
     this.rawMetadata = "loading";
-    const identityInfo = await this.connection.getIdentityInfo();
-    if ("Ok" in identityInfo) {
+    try {
       this.updatedMetadata = false;
-      this.rawMetadata = identityInfo.Ok.metadata;
-    } else {
+      this.rawMetadata = await this.getter();
+    } catch (error) {
+      // Do not throw the error becasue this is not critical for the application.
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:93]:
```suggestion
      // Do not throw the error because this is not critical for the application.
```

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+ * Class to manage the metadata of the anchor and interact with the canister.
+ *
+ * We decided to not throw any errors because this is non-critical for the application
+ * and we don't want to disrupt user flows if there is an error with the metadata.
+ *
+ * This class loads the metadata in the background when it's created.
+ * It can then be read and updated in memory.
+ * The metadata needs to be committed to the canister with the `commitMetadata` method to persist the changes.
+ */
+export class AnchorMetadataRepository {
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:44]: Quite a while ago we decided to sunset the term "anchor" because it is confusing. So on the UI, this term is no longer used. In the code, we didn't change things.
But for new code, we now use "identity" rather than anchor. Hence, why on the API level it is called `IdentityMetadata` rather than `AnchorMetadata`.
I think it would be better to use the new terminology here too and keep going with the slow journey of eventually replacing all the occurrences of "anchor".

File: src/frontend/src/utils/iiConnection.test.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    BigInt(1234),
+    mockActor
+  );
+
+  await vi.waitFor(() => expect(infoResponse).toEqual(mockRawMetadata));
+
+  expect(await connection.getAnchorMetadata()).toEqual(mockAnchorMetadata);
+});
+
+test("comits changes on anchor metadata", async () => {
```
Comment (frederikrothenberger) [inline src/frontend/src/utils/iiConnection.test.ts:74]:
```suggestion
test("commits changes on anchor metadata", async () => {
```

File: src/frontend/src/utils/iiConnection.test.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  get_anchor_info: vi.fn().mockResolvedValue(expectedAnchorInfo),
+  identity_info: vi.fn().mockImplementation(async () => {
+    // The `await` is necessary to make sure that the `getterResponse` is set before the test continues.
+    infoResponse = await mockRawMetadata;
+    return { Ok: { metadata: mockRawMetadata } };
+  }),
+  identity_metadata_replace: vi.fn().mockResolvedValue({ Ok: null }),
+} as unknown as ActorSubclass<_SERVICE>;
+
+test("gets anchor info from actor", async () => {
```
Comment (frederikrothenberger) [inline src/frontend/src/utils/iiConnection.test.ts:47]: While it is good to test things, I'm a bit confused as to how `getAnchorInfo` relates to the identity metadata functionality added in this PR?

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  // The nice AnchorMetadata is exposed to the outside world, while the raw metadata is kept private.
+  // We keep all the raw data to maintain other metadata fields.
+  private rawMetadata: RawMetadataState;
+  // Flag to keep track whether we need to commit the metadata to the canister.
+  private updatedMetadata: boolean;
+  private getter: MetadataGetter;
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:50]: These should be `readonly`?

File: src/frontend/src/repositories/anchorMetadata.test.ts [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+// Used to await that the getter has resolved.
+let getterResponse: MetadataMapV2 | null | undefined = null;
+const getterMockSuccess = vi.fn().mockImplementation(async () => {
+  // The `await` is necessary to make sure that the `getterResponse` is set before the test continues.
+  getterResponse = await mockRawMetadata;
+  return mockRawMetadata;
+});
+const getterMockError = vi.fn().mockImplementation(async () => {
+  // The `await` is necessary to make sure that the `getterResponse` is set before the test continues.
+  getterResponse = await null;
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.test.ts:28]: Sorry, I'm missing some context here. Why does the `await` change anything at all? Shouldn't that instantly resolve anyway?

File: src/frontend/src/repositories/anchorMetadata.ts [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+   *
+   * It tries to load the metadata if it's not loaded yet.
+   *
+   * It returns `undefined` if the metadata is not loaded.
+   *
+   * @returns {AnchorMetadata | undefined}
+   */
+  getMetadata = async (): Promise<AnchorMetadata | undefined> => {
+    if (!this.metadataIsLoaded(this.rawMetadata)) {
+      await this.loadMetadata();
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.ts:121]: Shouldn't we simply await `this.rawMetadata` here? Otherwise you might load things twice / are actually implementing an implicit retry?

Given it's not critical, we should probably avoid that, no?

File: src/frontend/src/repositories/anchorMetadata.test.ts [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+    setter: setterMockSuccess,
+  });
+
+  await vi.waitFor(() => expect(getterResponse).toEqual(mockRawMetadata));
+
+  expect(getterMockSuccess).toHaveBeenCalledTimes(1);
+  expect(await instance.getMetadata()).toEqual(mockAnchorMetadata);
+});
+
+test("AnchorMetadataRepository loads data again on init in the background", async () => {
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.test.ts:52]: See comment below, but I think the metadata should not be loaded _again_.

File: src/frontend/src/repositories/anchorMetadata.test.ts [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+
+  // Error is not thrown, but a warning is logged.
+  expect(console.warn).toHaveBeenCalledTimes(1);
+  expect(getterMockErrorOnce).toHaveBeenCalledTimes(1);
+
+  expect(await instance.getMetadata()).toEqual(mockAnchorMetadata);
+  expect(getterMockErrorOnce).toHaveBeenCalledTimes(2);
+});
+
+test("AnchorMetadataRepository returns no metadata if fetching fails without raising an error", async () => {
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.test.ts:78]:
```suggestion
test("AnchorMetadataRepository returns undefined without raising an error if fetching fails", async () => {
```

File: src/frontend/src/repositories/anchorMetadata.test.ts [hunk 16]
```diff
@@ -0,0 +1,223 @@
+import { MetadataMapV2 } from "$generated/internet_identity_types";
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/anchorMetadata.test.ts:1]: I think it would be good to have a test to ensure that the repository leaves additional data unchanged in `rawMetadata`.

File: src/frontend/src/utils/iiConnection.ts [hunk 17]
```diff
@@ -511,6 +538,34 @@ export class AuthenticatedConnection extends Connection {
...
+  ): Promise<{ Ok: null } | { Err: IdentityMetadataReplaceError }> => {
+    const actor = await this.getActor();
+    return await actor.identity_metadata_replace(this.userNumber, metadata);
+  };
+
+  getAnchorMetadata = (): Promise<AnchorMetadata | undefined> => {
+    return this.metadataRepository.getMetadata();
+  };
+
+  setPartialMetadata = async (
```
Comment (frederikrothenberger) [inline src/frontend/src/utils/iiConnection.ts:559]:
```suggestion
  updateIdentityMetadata = async (
```

**Review by frederikrothenberger — COMMENTED** [review]
Thanks! I still have a few questions.

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 7 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts]: Thanks!

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 8 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts]: Done!

File: src/frontend/src/utils/iiConnection.test.ts (hunk 9 above)
Comment (lmuntaner) [reply src/frontend/src/utils/iiConnection.test.ts]: Thanks!

File: src/frontend/src/utils/iiConnection.test.ts (hunk 10 above)
Comment (lmuntaner) [reply src/frontend/src/utils/iiConnection.test.ts]: I was trying that I could successfully test AuthenticatedConnection and then I thought it might be ok to leave it.

But let's do that in another PR if I want to add more tests. I removed it from this PR.

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 11 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts]: yes, done!

File: src/frontend/src/repositories/anchorMetadata.test.ts (hunk 12 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.test.ts]: For some reason it wasn't waiting before. But I tried again and it was working now. Removed.

File: src/frontend/src/repositories/anchorMetadata.ts (hunk 13 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.ts]: As agreed, there won't be a `loadMetadata` call again and this function just returns the pretty metadata.

File: src/frontend/src/repositories/anchorMetadata.test.ts (hunk 14 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.test.ts]: Done.

File: src/frontend/src/repositories/anchorMetadata.test.ts (hunk 15 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.test.ts]: Thanks!

File: src/frontend/src/repositories/anchorMetadata.test.ts (hunk 16 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/anchorMetadata.test.ts]: I had a test for this: `"IdentityMetadataRepository commits additional metadata to canister after update"`.

I rephrased the description to use "additional", maybe that was the confusing part.

File: src/frontend/src/utils/iiConnection.ts (hunk 17 above)
Comment (lmuntaner) [reply src/frontend/src/utils/iiConnection.ts:559]: done

**Comment by lmuntaner** [comment]
@frederikrothenberger ready for another review

**Comment by lmuntaner** [comment]
@frederikrothenberger wait a bit, I found a bug.

**Comment by lmuntaner** [comment]
@frederikrothenberger ready. I forgot to await for the rawMetadata.

File: src/frontend/src/repositories/identityMetadata.ts [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+   * This method won't throw an error if the metadata can't be loaded.
+   * Instead, it will set the instance variable `rawMetadata` to "error".
+   *
+   * @returns {Promise<void>} In case a client wants to wait for the metadata to be loaded.
+   */
+  loadMetadata = async (): Promise<void> => {
+    this.rawMetadata = "loading";
+    try {
+      this.updatedMetadata = false;
+      this.rawMetadata = await this.getter();
```
Comment (frederikrothenberger) [inline src/frontend/src/repositories/identityMetadata.ts:91]: You could also just put the `Promise` as a field. That way you could await it alongside a timer with `Promise.race(...)` in `getMetadata`, no?

Not sure about the error handling though.

**Review by frederikrothenberger — APPROVED** [review]
Thanks, lgtm! I left one suggestion but I don't feel strongly about it.

File: src/frontend/src/repositories/identityMetadata.ts (hunk 18 above)
Comment (lmuntaner) [reply src/frontend/src/repositories/identityMetadata.ts:91]: Yes, if I put the promise there I lose the states and I don't see the advantage. Unless I don't put the promise of the getter directly, but I wrap the getter so that it returns a promise of the states.

What do you think it's the advantage? I can do it in another PR. We would also lose the possibility of a client waiting on `loadMetadata`.
